File: app/src/main/java/com/example/despensacx/BackupHelper.java
package com.example.despensacx;

import android.content.Context;
import android.net.Uri;
import android.util.Log;

import com.google.gson.Gson;

import java.io.*;
import java.text.SimpleDateFormat;
import java.util.*;

public class BackupHelper {

    public static class BackupContainer {
        public List<ListaEntity> listas;
        public List<TiendaEntity> tiendas;
        public List<ProductoEntity> productos;
    }

    public static String generarNombreDefectoRespaldo() {
        String fecha = new SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(new Date());
        return "respaldo_" + fecha + ".json";
    }

    public static boolean exportarJSONToUri(Context context, Uri targetUri) {
        try (OutputStream os = context.getContentResolver().openOutputStream(targetUri);
             OutputStreamWriter writer = new OutputStreamWriter(os)) {

            AppDatabase db = AppDatabase.getInstance(context);
            BackupContainer container = new BackupContainer();
            container.listas = db.listaDao().getAllSync();
            container.tiendas = db.tiendaDao().getAllTiendasSync();
            container.productos = db.productoDao().getAllSync();

            String json = new Gson().toJson(container);
            writer.write(json);
            return true;
        } catch (Exception e) {
            Log.e("BackupHelper", "Error al exportar JSON vía SAF", e);
            return false;
        }
    }

    public static boolean importarJSONFromUri(Context context, Uri sourceUri) {
        try (InputStream is = context.getContentResolver().openInputStream(sourceUri);
             BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {

            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }

            BackupContainer container = new Gson().fromJson(sb.toString(), BackupContainer.class);
            if (container == null) return false;

            AppDatabase db = AppDatabase.getInstance(context);
            Map<Long, Long> mapaTiendasOldToNew = new HashMap<>();

            for (TiendaEntity t : container.tiendas) {
                long oldId = t.getId();
                t.setId(0);
                long newId = db.tiendaDao().insert(t);
                mapaTiendasOldToNew.put(oldId, newId);
            }

            for (ListaEntity l : container.listas) {
                long oldListaId = l.getId();
                l.setId(0);
                long newListaId = db.listaDao().insert(l);

                if (container.productos != null) {
                    for (ProductoEntity p : container.productos) {
                        if (p.getListaId() == oldListaId) {
                            p.setId(0);
                            p.setListaId(newListaId);
                            Long newTiendaId = mapaTiendasOldToNew.get(p.getTiendaId());
                            if (newTiendaId != null) {
                                p.setTiendaId(newTiendaId);
                                db.productoDao().insert(p);
                            }
                        }
                    }
                }
            }
            return true;
        } catch (Exception e) {
            Log.e("BackupHelper", "Error al importar JSON vía SAF", e);
            return false;
        }
    }
}

File: app/src/main/kotlin/com/example/despensacx/data/BackupHelper.kt
package com.example.despensacx.data

import android.content.Context
import android.net.Uri
import android.util.Log
import com.google.gson.Gson
import kotlinx.coroutines.runBlocking
import java.io.*
import java.util.*

object BackupHelper {

    class BackupContainer {
        var listas: List<ListaEntity>? = null
        var tiendas: List<TiendaEntity>? = null
        var productos: List<ProductoEntity>? = null
        var catalogo: List<CatalogoProducto>? = null
        var membresias: List<MembresiaEntity>? = null
    }

    fun generarNombreDefectoRespaldo(): String {
        return "DespensaCX_Backup.json"
    }

    fun exportarJSONToUri(context: Context, targetUri: Uri): Boolean {
        return try {
            context.contentResolver.openOutputStream(targetUri)?.use { os ->
                OutputStreamWriter(os).use { writer ->
                    val db = AppDatabase.getInstance(context)
                    val container = BackupContainer().apply {
                        listas = db.listaDao().getAllSync()
                        tiendas = db.tiendaDao().getAllTiendasSync()
                        productos = db.productoDao().getAllSync()
                        catalogo = db.catalogoDao().getAllSync()
                    }
                    val json = Gson().toJson(container)
                    writer.write(json)
                    true
                }
            } ?: false
        } catch (e: Exception) {
            Log.e("BackupHelper", "Error al exportar JSON", e)
            false
        }
    }

    fun importarJSONFromUri(context: Context, sourceUri: Uri): Boolean {
        return try {
            context.contentResolver.openInputStream(sourceUri)?.use { `is` ->
                importarJSONFromInputStream(context, `is`)
            } ?: false
        } catch (e: Exception) {
            Log.e("BackupHelper", "Error al importar JSON", e)
            false
        }
    }

    fun importarJSONFromInputStream(context: Context, inputStream: InputStream): Boolean {
        return try {
            val reader = BufferedReader(InputStreamReader(inputStream))
            val sb = StringBuilder()
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                sb.append(line)
            }

            val container = Gson().fromJson(sb.toString(), BackupContainer::class.java) ?: return false
            val db = AppDatabase.getInstance(context)
            val mapaTiendasOldToNew = mutableMapOf<Long, Long>()

            runBlocking {
                db.productoDao().deleteAll()
                db.listaDao().deleteAll()
                db.tiendaDao().deleteAll()
                db.catalogoDao().deleteAll()
                db.ticketFotoDao().deleteAll()
                db.membresiaDao().deleteAll()

                container.catalogo?.forEach { db.catalogoDao().insert(it) }
                container.tiendas?.forEach { t ->
                    val oldId = t.id
                    t.id = 0
                    mapaTiendasOldToNew[oldId] = db.tiendaDao().insert(t)
                }

                container.membresias?.forEach { m ->
                    val newTiendaId = mapaTiendasOldToNew[m.tiendaId]
                    if (newTiendaId != null) {
                        db.membresiaDao().insert(m.copy(id = 0, tiendaId = newTiendaId))
                    }
                }

                container.listas?.forEach { l ->
                    val oldListaId = l.id
                    l.id = 0
                    val newListaId = db.listaDao().insert(l)

                    container.productos?.filter { it.listaId == oldListaId }?.forEach { p ->
                        p.id = 0
                        p.listaId = newListaId
                        p.tiendaId = mapaTiendasOldToNew[p.tiendaId] ?: p.tiendaId
                        db.productoDao().insert(p)
                    }
                }
            }
            true
        } catch (e: Exception) {
            Log.e("BackupHelper", "Error procesando JSON", e)
            false
        }
    }
}
